Tab context action

pinussilvestrus opened the pull request:
Closes #2630 

This includes
* firing an `app.tabsChanged` event whenever the tabs change
* adding a `maxHeight` prop for overlay dropdown option groups
* adding the tab context action (requirements cf. #2630)

![image](https://user-images.githubusercontent.com/9433996/145832678-fc0ff323-f6f8-4e85-b0a2-4ce800757757.png)

![image](https://user-images.githubusercontent.com/9433996/145832863-bb72301a-161f-4cd7-8c88-30e8dadb830d.png)


------

Artifacts are ready to use 
* https://camunda-modeler-on-demand.s3.eu-central-1.amazonaws.com/2630-tab-context-menu/camunda-modeler-2630-tab-context-menu-linux-x64.tar.gz
* https://camunda-modeler-on-demand.s3.eu-central-1.amazonaws.com/2630-tab-context-menu/camunda-modeler-2630-tab-context-menu-mac.dmg
* https://camunda-modeler-on-demand.s3.eu-central-1.amazonaws.com/2630-tab-context-menu/camunda-modeler-2630-tab-context-menu-mac.zip
* https://camunda-modeler-on-demand.s3.eu-central-1.amazonaws.com/2630-tab-context-menu/camunda-modeler-2630-tab-context-menu-win-ia32.zip
* https://camunda-modeler-on-demand.s3.eu-central-1.amazonaws.com/2630-tab-context-menu/camunda-modeler-2630-tab-context-menu-win-x64.zip

andreasgeier commented:
Looks great already! Two topics:

(1) Can you avoid to show "Save all" in case there is nothing to save?

(2) Please change overlay max-height from `auto` to `80vh`, it should not "dock" at the bottom but have some space. Expected result:

![Screen Shot 2021-12-14 at 11 52 26](https://user-images.githubusercontent.com/938244/145985077-121351e4-2bc9-4256-8436-dee07ebeba0e.png)

nikku commented:
> (1) Can you avoid to show "Save all" in case there is nothing to save? It should be available in case (a) there is at least one newly created file or (b) there is at least one file with dirty marker

I'd like to not go that route. For me "save all" is a no-regret move, which defaults to no-op, if nothing needs to be saved. It would confuse the hell out of me as a user to not have it available some day.

Also double checked against my IDE to verify that this is not a direction it goes.

andreasgeier commented:
> I'd like to not go that route. For me "save all" is a no-regret move, which defaults to no-op, if nothing needs to be saved. It would confuse the hell out of me as a user to not have it available if I wanted to simply ensure "everything is saved".

This follows our basic rule: "don't show actions that are not available". See the 3 close actions, they are only available if one (or multiple) tabs are opened. I want to stick to that rule and avoid a button that can be triggered but wouldn't do anything.

Btw. not having an active save button is a no-worries indicator, too.

pinussilvestrus commented:
I'd agree with Nico on this point, although given we have the rule that we don't show actions that are not available. In this case, the action is actually doing something, namely

>  ensure "everything is saved"

no matter what. The dirty marker doesn't give me enough indicator that everything is saved, since new files don't have it without any changes, but would need to be saved anyway.

But, it seems like we have different opinions here. Maybe then just keep things as they are (always displaying the save action) and wait for feedback?

andreasgeier commented:
I would like to make a decision here to stick to the rule. In addition to the arguments that I already mentioned it would cause inconsistency with the app menu. There, "Save" is greyed out in case there is nothing to save.

pinussilvestrus commented:
For the sake of being consistent with the application menu, I'd agree - although I still believe "saving all" is always doing something, maybe not technically, but at least in giving the feel of "everything is saved". As an alternative: we could also say, let's always enable "save all", in the context action and also in the application menu.

However, if we decide to go with removing as @andreasgeier suggested, I added https://github.com/camunda/camunda-modeler/pull/2631/commits/cca378313ca64f70fccc4701896dded05d85ff28 on top. 

I also adjusted the max-height as suggested.

nikku commented:
@andreasgeier Greying stuff out is not removing things. And save all does still work, it just does nothing. There is nothing inconsistent of keeping it active in the menu, from my point of view.

andreasgeier reviewed (approved):
Thanks, @pinussilvestrus for removing the save button in case it is not needed.

Let's not make a big discussion about this. Having a non-functioning button in the UI which is active but does nothing is just bad UX practice. And I would counter the speculation that users need a save button "just in case of uncertainty". My assumption is that almost every desktop user knows that it is not possible to close unsaved files without a warning dialog.

✅ &nbsp; Approved

nikku commented:
> My assumption is that almost every desktop user knows that it is not possible to close unsaved files without a warning dialog.

This is not about "closing" but about the guarantee that everything is saved at certain checkpoints (I'm low on battery, have a computer that crashes from time to time). Automatic safe does solve this in a different manner, but introduces other drawbacks.

MaxTru commented:
We discussed the `Save all` incl. Design, Engineering, and PM. To conclude the disagreement, as interim PM I decided to have the `Save all` included the same way we have it in the file menu (ie. always show it, potentially be a noop). This will make it consistent. We might iterate in the future, but not for the initial 5.0.0 release.

MaxTru reviewed (changes requested):
See my comment, let's please include the Save all as decided

pinussilvestrus commented:
Thanks for driving a decision here! I've removed the respective commit (https://github.com/camunda/camunda-modeler/commit/cca378313ca64f70fccc4701896dded05d85ff28) accordingly.

MaxTru commented on client/src/app/__tests__/AppSpec.js:
```diff
@@ -735,6 +735,27 @@ describe('<App>', function() {
     });
 
+
+    it('should emit <app.tabsChanged> event on tab closed', async function() {
```
Should this also emit when we change a tab? (ie. click on an already opened tab?) In this case, we miss a test case I believe?

Currently we only test:
* tab close
* tab create

pinussilvestrus replied:
👍  I added a test case via https://github.com/camunda/camunda-modeler/pull/2631/commits/189ef000355015fa9a7d75440639a79e7992535c accordingly.

MaxTru reviewed (approved):
Nice!
